File: packages/server-python/app/routers/gameplay.py
"""Memory Match gameplay routes."""

import random
from datetime import datetime
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.game import GamePlayer, GameRoom, GameState
from app.models.user import User
from app.schemas.gameplay import FlipCardRequest

router = APIRouter(prefix="/gameplay", tags=["gameplay"])

# Emoji symbols for the memory game
SYMBOLS = ["🎮", "🎯", "🎲", "🎪", "🎨", "🎭", "🎬", "🎸", "🎹", "🎤", "🎧", "🎼"]


def create_memory_deck(pairs: int = 8) -> list[dict]:
    """Create a shuffled deck of memory cards."""
    selected_symbols = SYMBOLS[:pairs]
    cards = []

    for idx, symbol in enumerate(selected_symbols):
        cards.append({"id": idx * 2, "value": symbol, "isFlipped": False, "isMatched": False})
        cards.append({"id": idx * 2 + 1, "value": symbol, "isFlipped": False, "isMatched": False})

    random.shuffle(cards)
    return cards


@router.get("/{room_id}/state")
def get_game_state(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Get the current game state."""
    # Verify player is in the room
    player = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id, GamePlayer.user_id == user.id)
        .first()
    )
    if not player:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not in this room",
        )

    # Get game state
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if not state_row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not started yet",
        )

    # Get room
    room = db.query(GameRoom).filter(GameRoom.id == room_id).first()

    # Get all players
    players = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id)
        .order_by(GamePlayer.position)
        .all()
    )

    return {
        "room": {
            "id": room.id,
            "gameId": room.game_id,
            "name": room.name,
            "hostId": room.host_id,
            "status": room.status,
            "maxPlayers": room.max_players,
        },
        "state": state_row.state,
        "players": [
            {"id": p.id, "userId": p.user_id, "position": p.position, "score": p.score}
            for p in players
        ],
        "currentPlayerPosition": player.position,
    }


@router.post("/{room_id}/initialize")
def initialize_game(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Initialize the memory match game (host only)."""
    room = db.query(GameRoom).filter(GameRoom.id == room_id).first()
    if not room:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Room not found",
        )

    if room.host_id != user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only host can initialize the game",
        )

    # Get players
    players = db.query(GamePlayer).filter(GamePlayer.room_id == room_id).all()

    # Initialize scores
    scores = {str(p.position): 0 for p in players}

    # Create initial state
    initial_state = {
        "cards": create_memory_deck(8),
        "flippedIndices": [],
        "matchedPairs": 0,
        "currentPlayerPosition": 0,
        "scores": scores,
    }

    # Upsert game state
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if state_row:
        state_row.state = initial_state
        state_row.current_turn = 0
        state_row.updated_at = datetime.utcnow()
    else:
        state_row = GameState(room_id=room_id, current_turn=0)
        state_row.state = initial_state
        db.add(state_row)

    db.commit()
    return {"success": True}


@router.post("/{room_id}/flip")
def flip_card(
    room_id: str,
    body: FlipCardRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Flip a card."""
    # Verify player is in the room
    player = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id, GamePlayer.user_id == user.id)
        .first()
    )
    if not player:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not in this room",
        )

    # Get game state
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if not state_row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not started",
        )

    state = state_row.state

    # Check if it's this player's turn
    if state["currentPlayerPosition"] != player.position:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not your turn",
        )

    # Check if already 2 cards flipped
    if len(state["flippedIndices"]) >= 2:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Please reset flipped cards first",
        )

    # Find card index
    card_index = next(
        (i for i, c in enumerate(state["cards"]) if c["id"] == body.cardId), None
    )
    if card_index is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Card not found",
        )

    card = state["cards"][card_index]
    if card["isMatched"] or card["isFlipped"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Card already flipped or matched",
        )

    # Flip the card
    state["cards"][card_index]["isFlipped"] = True
    state["flippedIndices"].append(card_index)

    # Check for match if 2 cards flipped
    if len(state["flippedIndices"]) == 2:
        idx1, idx2 = state["flippedIndices"]
        card1 = state["cards"][idx1]
        card2 = state["cards"][idx2]

        if card1["value"] == card2["value"]:
            # Match found
            state["cards"][idx1]["isMatched"] = True
            state["cards"][idx2]["isMatched"] = True
            state["matchedPairs"] += 1
            state["scores"][str(player.position)] = (
                state["scores"].get(str(player.position), 0) + 1
            )

            # Update player score in database
            player.score = state["scores"][str(player.position)]

            # Check if game is finished
            if state["matchedPairs"] == len(state["cards"]) // 2:
                room = db.query(GameRoom).filter(GameRoom.id == room_id).first()
                room.status = "finished"
                room.finished_at = datetime.utcnow()

    # Save state
    state_row.state = state
    state_row.updated_at = datetime.utcnow()
    db.commit()

    return {"success": True, "state": state}


@router.post("/{room_id}/reset")
def reset_flipped(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Reset unmatched flipped cards and advance turn."""
    # Verify player is in the room
    player = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id, GamePlayer.user_id == user.id)
        .first()
    )
    if not player:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not in this room",
        )

    # Get game state
    state_row = db.query(GameState).filter(GameState.room_id == room_id).first()
    if not state_row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not started",
        )

    state = state_row.state

    # Check if it's this player's turn
    if state["currentPlayerPosition"] != player.position:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Not your turn",
        )

    # If 2 cards flipped and not matched, flip them back and advance turn
    if len(state["flippedIndices"]) == 2:
        idx1, idx2 = state["flippedIndices"]
        card1 = state["cards"][idx1]
        card2 = state["cards"][idx2]

        if card1["value"] != card2["value"]:
            state["cards"][idx1]["isFlipped"] = False
            state["cards"][idx2]["isFlipped"] = False

            # Get player count
            player_count = (
                db.query(GamePlayer).filter(GamePlayer.room_id == room_id).count()
            )
            state["currentPlayerPosition"] = (
                state["currentPlayerPosition"] + 1
            ) % player_count

    # Clear flipped indices
    state["flippedIndices"] = []

    # Save state
    state_row.state = state
    state_row.updated_at = datetime.utcnow()
    db.commit()

    return {"success": True, "state": state}

File: src/server/routes/gameplay.ts
import { ApiRouter, t } from '../api-router'
import { db } from '../db/index.js'
import { gameRooms, gamePlayers, gameState } from '../db/schema.js'
import { eq, and } from 'drizzle-orm'
import { authMiddleware } from '../middleware/auth.js'

export interface MemoryCard {
  id: number
  value: string
  isFlipped: boolean
  isMatched: boolean
}

export interface MemoryMatchState {
  cards: MemoryCard[]
  flippedIndices: number[]
  matchedPairs: number
  currentPlayerPosition: number
  scores: { [position: number]: number }
}

function createMemoryMatchDeck(pairs: number = 8): MemoryCard[] {
  const symbols = ['🎮', '🎯', '🎲', '🎪', '🎨', '🎭', '🎬', '🎸', '🎹', '🎤', '🎧', '🎼']
  const selectedSymbols = symbols.slice(0, pairs)
  const cards: MemoryCard[] = []

  selectedSymbols.forEach((symbol, idx) => {
    cards.push(
      {
        id: idx * 2,
        value: symbol,
        isFlipped: false,
        isMatched: false,
      },
      {
        id: idx * 2 + 1,
        value: symbol,
        isFlipped: false,
        isMatched: false,
      },
    )
  })

  for (let i = cards.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[cards[i], cards[j]] = [cards[j], cards[i]]
  }

  return cards
}

export const gameplayRoutes = new ApiRouter({ prefix: '/api/gameplay' })
  .use(authMiddleware)
  .get(
    '/:roomId/state',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const player = await db
        .select()
        .from(gamePlayers)
        .where(and(eq(gamePlayers.roomId, params.roomId), eq(gamePlayers.userId, user.id)))
        .limit(1)

      if (!player[0]) {
        set.status = 403
        return { error: 'Forbidden', message: 'You are not in this room' }
      }

      const state = await db
        .select()
        .from(gameState)
        .where(eq(gameState.roomId, params.roomId))
        .limit(1)

      if (!state[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not started yet' }
      }

      const room = await db
        .select()
        .from(gameRooms)
        .where(eq(gameRooms.id, params.roomId))
        .limit(1)

      const players = await db
        .select()
        .from(gamePlayers)
        .where(eq(gamePlayers.roomId, params.roomId))
        .orderBy(gamePlayers.position)

      return {
        room: room[0],
        state: state[0].state as MemoryMatchState,
        players,
        currentPlayerPosition: player[0].position,
      }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .post(
    '/:roomId/initialize',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const room = await db
        .select()
        .from(gameRooms)
        .where(eq(gameRooms.id, params.roomId))
        .limit(1)

      if (!room[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Room not found' }
      }

      if (room[0].hostId !== user.id) {
        set.status = 403
        return { error: 'Forbidden', message: 'Only host can initialize the game' }
      }

      const players = await db.select().from(gamePlayers).where(eq(gamePlayers.roomId, params.roomId))

      const scores: { [position: number]: number } = {}
      players.forEach((player) => {
        scores[player.position] = 0
      })

      const initialState: MemoryMatchState = {
        cards: createMemoryMatchDeck(8),
        flippedIndices: [],
        matchedPairs: 0,
        currentPlayerPosition: 0,
        scores,
      }

      await db
        .insert(gameState)
        .values({
          roomId: params.roomId,
          state: initialState,
          currentTurn: 0,
        })
        .onConflictDoUpdate({
          target: gameState.roomId,
          set: {
            state: initialState,
            currentTurn: 0,
            updatedAt: new Date(),
          },
        })

      return { success: true }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .post(
    '/:roomId/flip',
    async ({ params, body, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const player = await db
        .select()
        .from(gamePlayers)
        .where(and(eq(gamePlayers.roomId, params.roomId), eq(gamePlayers.userId, user.id)))
        .limit(1)

      if (!player[0]) {
        set.status = 403
        return { error: 'Forbidden', message: 'You are not in this room' }
      }

      const stateRow = await db
        .select()
        .from(gameState)
        .where(eq(gameState.roomId, params.roomId))
        .limit(1)

      if (!stateRow[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not started' }
      }

      const state = stateRow[0].state as MemoryMatchState

      if (state.currentPlayerPosition !== player[0].position) {
        set.status = 400
        return { error: 'Bad request', message: 'Not your turn' }
      }

      if (state.flippedIndices.length >= 2) {
        set.status = 400
        return { error: 'Bad request', message: 'Please reset flipped cards first' }
      }

      const cardIndex = state.cards.findIndex((c) => c.id === body.cardId)
      if (cardIndex === -1) {
        set.status = 404
        return { error: 'Not found', message: 'Card not found' }
      }

      const card = state.cards[cardIndex]
      if (card.isMatched || card.isFlipped) {
        set.status = 400
        return { error: 'Bad request', message: 'Card already flipped or matched' }
      }

      state.cards[cardIndex].isFlipped = true
      state.flippedIndices.push(cardIndex)

      if (state.flippedIndices.length === 2) {
        const [idx1, idx2] = state.flippedIndices
        const card1 = state.cards[idx1]
        const card2 = state.cards[idx2]

        if (card1.value === card2.value) {
          state.cards[idx1].isMatched = true
          state.cards[idx2].isMatched = true
          state.matchedPairs++
          state.scores[player[0].position]++

          await db
            .update(gamePlayers)
            .set({
              score: state.scores[player[0].position],
            })
            .where(eq(gamePlayers.id, player[0].id))

          if (state.matchedPairs === state.cards.length / 2) {
            await db
              .update(gameRooms)
              .set({
                status: 'finished',
                finishedAt: new Date(),
              })
              .where(eq(gameRooms.id, params.roomId))
          }
        }
      }

      await db
        .update(gameState)
        .set({
          state,
          updatedAt: new Date(),
        })
        .where(eq(gameState.roomId, params.roomId))

      return { success: true, state }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
      body: t.Object({
        cardId: t.Number(),
      }),
    },
  )
  .post(
    '/:roomId/reset',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const player = await db
        .select()
        .from(gamePlayers)
        .where(and(eq(gamePlayers.roomId, params.roomId), eq(gamePlayers.userId, user.id)))
        .limit(1)

      if (!player[0]) {
        set.status = 403
        return { error: 'Forbidden', message: 'You are not in this room' }
      }

      const stateRow = await db
        .select()
        .from(gameState)
        .where(eq(gameState.roomId, params.roomId))
        .limit(1)

      if (!stateRow[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not started' }
      }

      const state = stateRow[0].state as MemoryMatchState

      if (state.currentPlayerPosition !== player[0].position) {
        set.status = 400
        return { error: 'Bad request', message: 'Not your turn' }
      }

      if (state.flippedIndices.length === 2) {
        const [idx1, idx2] = state.flippedIndices
        const card1 = state.cards[idx1]
        const card2 = state.cards[idx2]

        if (card1.value !== card2.value) {
          state.cards[idx1].isFlipped = false
          state.cards[idx2].isFlipped = false

          const players = await db
            .select()
            .from(gamePlayers)
            .where(eq(gamePlayers.roomId, params.roomId))

          state.currentPlayerPosition = (state.currentPlayerPosition + 1) % players.length
        }
      }

      state.flippedIndices = []

      await db
        .update(gameState)
        .set({
          state,
          updatedAt: new Date(),
        })
        .where(eq(gameState.roomId, params.roomId))

      return { success: true, state }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
